crates/oxc_linter/src/rules/typescript/no_restricted_types.rs

```diff
@@ -0,0 +1,558 @@
+    Bool(True),
+    /// `"TypeName": "message"` - ban with custom message
+    Message(String),
+    /// `"TypeName": { "message": "...", "fixWith": "...", "suggest": ["..."] }` - full config
+    Object {
+        #[serde(default)]
+        message: Option<String>,
+        #[serde(default, rename = "fixWith")]
+        fix_with: Option<String>,
+        #[serde(default)]
+        suggest: Option<Vec<String>>,
+    },
+}
+
+/// A type that only deserializes from `true`.
+/// This matches the upstream typescript-eslint schema which only allows `true`, not `false`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, JsonSchema)]
+struct True;
+
+impl<'de> Deserialize<'de> for True {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value = bool::deserialize(deserializer)?;
+        if value { Ok(True) } else { Err(serde::de::Error::custom("expected `true`, got `false`")) }
+    }
+}
+
+impl BanConfigValue {
+    fn message(&self) -> Option<&str> {
+        match self {
+            BanConfigValue::Bool(_) => None,
+            BanConfigValue::Message(msg) => Some(msg.as_str()),
+            BanConfigValue::Object { message, .. } => message.as_deref(),
+        }
+    }
+
+    fn fix_with(&self) -> Option<&str> {
+        match self {
+            BanConfigValue::Bool(_) | BanConfigValue::Message(_) => None,
+            BanConfigValue::Object { fix_with, .. } => fix_with.as_deref(),
+        }
+    }
+
+    fn suggest(&self) -> Option<&[String]> {
+        match self {
+            BanConfigValue::Bool(_) | BanConfigValue::Message(_) => None,
+            BanConfigValue::Object { suggest, .. } => suggest.as_deref(),
+        }
+    }
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallow certain types from being used.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Some built-in types have aliases, while some types are considered dangerous or harmful.
+    /// It's often a good idea to ban certain types to help with consistency and safety.
+    ///
+    /// ### Examples
+    ///
+    /// Given `{ "types": { "Foo": { "message": "Use Bar instead", "fixWith": "Bar" } } }`:
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```ts
+    /// let value: Foo;
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```ts
+    /// let value: Bar;
+    /// ```
+    NoRestrictedTypes,
+    typescript,
+    restriction,
+    fix_suggestion,
+    config = NoRestrictedTypesConfig,
+);
+
+impl Rule for NoRestrictedTypes {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let config = value
+            .get(0)
+            .and_then(|v| serde_json::from_value::<NoRestrictedTypesConfig>(v.clone()).ok())
+            .unwrap_or_default();
+        Self(Box::new(config))
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
```

Comment:
Noob question so possibly irrelevant, but: Is this matcher sufficiently performant? e.g. should we do anything to pre-calculate which ast kinds to check based on the banned types, rather than doing that check every time there's a node of any of these types?

Reply:
Also, would this automatically skip these checks for an empty array of banned types? And can this be implemented as `run_once`?